<a href="https://colab.research.google.com/github/dean-sh/Movie-Ratings-Collaborating-Filltering/blob/master/Singular%20Value%20Decomposition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

MovieLens Recommendations - Dean Shabi, Dedi Kovatch, July 2019
=============================================

## Final Project for TCDS - Technion Data Science Specialization. 

MovieLens data sets were collected by the GroupLens Research Project
at the University of Minnesota.
 
This data set consists of:
	* 100,000 ratings (1-5) from 943 users on 1682 movies. 
	* Each user has rated at least 20 movies. 
        * Simple demographic info for the users (age, gender, occupation, zip)

The data was collected through the MovieLens web site
(movielens.umn.edu) during the seven-month period from September 19th, 
1997 through April 22nd, 1998. This data has been cleaned up - users
who had less than 20 ratings or did not have complete demographic
information were removed from this data set. Detailed descriptions of
the data file can be found at the end of this file.


## Data Description




Here are brief descriptions of the data.

ml-data.tar.gz   -- Compressed tar file.  To rebuild the u data files do this:
                gunzip ml-data.tar.gz
                tar xvf ml-data.tar
                mku.sh

u.data     -- The full u data set, 100000 ratings by 943 users on 1682 items.
              Each user has rated at least 20 movies.  Users and items are
              numbered consecutively from 1.  The data is randomly
              ordered. This is a tab separated list of 
	         user id | item id | rating | timestamp. 
              The time stamps are unix seconds since 1/1/1970 UTC   

u.info     -- The number of users, items, and ratings in the u data set.

u.item     -- Information about the items (movies); this is a tab separated
              list of
              movie id | movie title | release date | video release date |
              IMDb URL | unknown | Action | Adventure | Animation |
              Children's | Comedy | Crime | Documentary | Drama | Fantasy |
              Film-Noir | Horror | Musical | Mystery | Romance | Sci-Fi |
              Thriller | War | Western |
              The last 19 fields are the genres, a 1 indicates the movie
              is of that genre, a 0 indicates it is not; movies can be in
              several genres at once.
              The movie ids are the ones used in the u.data data set.

u.genre    -- A list of the genres.

u.user     -- Demographic information about the users; this is a tab
              separated list of
              user id | age | gender | occupation | zip code
              The user ids are the ones used in the u.data data set.

u.occupation -- A list of the occupations.

u1.base    -- The data sets u1.base and u1.test through u5.base and u5.test
u1.test       are 80%/20% splits of the u data into training and test data.
u2.base       Each of u1, ..., u5 have disjoint test sets; this if for
u2.test       5 fold cross validation (where you repeat your experiment
u3.base       with each training and test set and average the results).
u3.test       These data sets can be generated from u.data by mku.sh.
u4.base
u4.test
u5.base
u5.test

ua.base    -- The data sets ua.base, ua.test, ub.base, and ub.test
ua.test       split the u data into a training set and a test set with
ub.base       exactly 10 ratings per user in the test set.  The sets
ub.test       ua.test and ub.test are disjoint.  These data sets can
              be generated from u.data by mku.sh.

allbut.pl  -- The script that generates training and test sets where
              all but n of a users ratings are in the training data.

mku.sh     -- A shell script to generate all the u data sets from u.data.

## Imports




In [0]:
import numpy as np
import pandas as pd
import collections
import seaborn as sns
%matplotlib inline
# from mpl_toolkits.mplot3d import Axes3D
from IPython import display
from matplotlib import pyplot as plt
import sklearn
import sklearn.manifold
# import tensorflow as tf
# tf.logging.set_verbosity(tf.logging.ERROR)

# # Add some convenience functions to Pandas DataFrame.
# pd.options.display.max_rows = 10
# pd.options.display.float_format = '{:.3f}'.format



# # Install Altair and activate its colab renderer.
# print("Installing Altair...")
# !pip install git+git://github.com/altair-viz/altair.git
# import altair as alt
# alt.data_transformers.enable('default', max_rows=None)
# alt.renderers.enable('colab')
# print("Done installing Altair.")

# # Install spreadsheets and import authentication module.
# USER_RATINGS = False
# !pip install --upgrade -q gspread
# from google.colab import auth
# import gspread
# from oauth2client.client import GoogleCredentials

## **Importing dataset, preprocessing**




In [0]:
# download the MovieLens Data, and create DataFrames containing movies, users, and ratings.

print("Downloading movielens data...")
import zipfile
import urllib.request

urllib.request.urlretrieve("http://files.grouplens.org/datasets/movielens/ml-100k.zip", "movielens.zip")
zip_ref = zipfile.ZipFile('movielens.zip', "r")
zip_ref.extractall()
print("Done. Dataset contains:")
print(zip_ref.read('ml-100k/u.info'))

In [0]:
# Load each data set (users, movies, and ratings).
users_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']
users = pd.read_csv(
    'ml-100k/u.user', sep='|', names=users_cols, encoding='latin-1')

ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv(
    'ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')

# The movies file contains a binary feature for each genre.
genre_cols = [
    "genre_unknown", "Action", "Adventure", "Animation", "Children", "Comedy",
    "Crime", "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror",
    "Musical", "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"
]
movies_cols = [
    'movie_id', 'title', 'release_date', "video_release_date", "imdb_url"] + genre_cols

movies = pd.read_csv(
    'ml-100k/u.item', sep='|', names=movies_cols, encoding='latin-1')


Some Preproccessing

In [0]:
# Since the ids start at 1, we shift them to start at 0.
users["user_id"] = users["user_id"].apply(lambda x: str(x-1))
movies["movie_id"] = movies["movie_id"].apply(lambda x: str(x-1))
movies["year"] = movies['release_date'].apply(lambda x: str(x).split('-')[-1])
ratings["movie_id"] = ratings["movie_id"].apply(lambda x: str(x-1))
ratings["user_id"] = ratings["user_id"].apply(lambda x: str(x-1))
ratings["rating"] = ratings["rating"].apply(lambda x: float(x))

# Compute the number of movies to which a genre is assigned.
genre_occurences = movies[genre_cols].sum().to_dict()

# Since some movies can belong to more than one genre, we create different
# 'genre' columns as follows:
# - all_genres: all the active genres of the movie.
# - genre: randomly sampled from the active genres.
def mark_genres(movies, genres):
    def get_random_genre(gs):
        active = [genre for genre, g in zip(genres, gs) if g==1]
        if len(active) == 0:
            return 'Other'
        return np.random.choice(active)
    def get_all_genres(gs):
        active = [genre for genre, g in zip(genres, gs) if g==1]
        if len(active) == 0:
            return 'Other'
        return '-'.join(active)
    movies['genre'] = [
        get_random_genre(gs) for gs in zip(*[movies[genre] for genre in genres])]
    movies['all_genres'] = [
        get_all_genres(gs) for gs in zip(*[movies[genre] for genre in genres])]

mark_genres(movies, genre_cols)

In [0]:
# Create one merged DataFrame containing all the movielens data.
movielens = ratings.merge(movies, on='movie_id').merge(users, on='user_id')

# Utility to split the data into training and test sets.
def split_dataframe(df, holdout_fraction=0.1):
    """Splits a DataFrame into training and test sets.
    Args:
    df: a dataframe.
    holdout_fraction: fraction of dataframe rows to use in the test set.
    Returns:
    train: dataframe for training
    test: dataframe for testing
    """
    test = df.sample(frac=holdout_fraction, replace=False)
    train = df[~df.index.isin(test.index)]
    return train, test

In [0]:
movies.head()

movie_id              title  ...      genre                 all_genres
0        0   Toy Story (1995)  ...     Comedy  Animation-Children-Comedy
1        1   GoldenEye (1995)  ...  Adventure  Action-Adventure-Thriller
2        2  Four Rooms (1995)  ...   Thriller                   Thriller
3        3  Get Shorty (1995)  ...     Comedy        Action-Comedy-Drama
4        4     Copycat (1995)  ...      Crime       Crime-Drama-Thriller

[5 rows x 27 columns]

In [0]:
users.head()

user_id  age sex  occupation zip_code
0       0   24   M  technician    85711
1       1   53   F       other    94043
2       2   23   M      writer    32067
3       3   24   M  technician    43537
4       4   33   F       other    15213

In [0]:
ratings.head()

user_id movie_id  rating  unix_timestamp
0     195      241     3.0       881250949
1     185      301     3.0       891717742
2      21      376     1.0       878887116
3     243       50     2.0       880606923
4     165      345     1.0       886397596

In [0]:
movielens.head()

user_id movie_id  rating  unix_timestamp  ... age sex  occupation zip_code
0     195      241     3.0       881250949  ...  49   M      writer    55105
1     195      256     2.0       881251577  ...  49   M      writer    55105
2     195      110     4.0       881251793  ...  49   M      writer    55105
3     195       24     4.0       881251955  ...  49   M      writer    55105
4     195      381     4.0       881251843  ...  49   M      writer    55105

[5 rows x 34 columns]

# Matrix Factorization via Singular Value Decomposition

Based on great work done by **Nick Becker**, RAPIDS Team at NVIDIA

*   https://beckernick.github.io/matrix-factorization-recommender/
*   https://github.com/beckernick/matrix_factorization_recommenders

This gave us great recommendations, however the RMSE we got was pretty high (probably due to some normalization factor we didn't account for)


Matrix factorization is the breaking down of one matrix in a product of multiple matrices. It's extremely well studied in mathematics, and it's highly useful. There are many different ways to factor matrices, but singular value decomposition is particularly useful for making recommendations.

So what is singular value decomposition (SVD)? At a high level, SVD is an algorithm that decomposes a matrix $R$ into the best lower rank (i.e. smaller/simpler) approximation of the original matrix $R$. Mathematically, it decomposes R into a two unitary matrices and a diagonal matrix:

$$\begin{equation}
R = U\Sigma V^{T}
\end{equation}$$

where R is users's ratings matrix, $U$ is the user "features" matrix, $\Sigma$ is the diagonal matrix of singular values (essentially weights), and $V^{T}$ is the movie "features" matrix. $U$ and $V^{T}$ are orthogonal, and represent different things. $U$ represents how much users "like" each feature and $V^{T}$ represents how relevant each feature is to each movie.

To get the lower rank approximation, we take these matrices and keep only the top $k$ features, which we think of as the underlying tastes and preferences vectors.


## Setting Up the Ratings Data

In [0]:
import pandas as pd
import numpy as np
import zipfile
import urllib.request

print("Downloading movielens data...")
    
urllib.request.urlretrieve("http://files.grouplens.org/datasets/movielens/ml-latest-small.zip", "movielens.zip")

zip_ref = zipfile.ZipFile('movielens.zip', "r")
zip_ref.extractall()


ratings_df = pd.read_csv('ml-latest-small/ratings.csv', names=['user_id', 'movie_id', 'rating', 'timestamp'], sep=',', encoding='latin-1', header = None)
ratings_df.drop([0], inplace=True)
ratings_df=ratings_df.apply(pd.to_numeric)
# ratings_df['UserID'] = ratings_df['UserID'].apply(pd.to_numeric)
# ratings_df['UserID'] = ratings_df['UserID'].apply(pd.to_numeric)


movies_df = pd.read_csv('ml-latest-small/movies.csv',names= ['movie_id', 'title', 'genres'], sep=',', encoding='latin-1')
movies_df.drop([0], inplace=True)
movies_df['movie_id'] = movies_df['movie_id'].apply(pd.to_numeric)
# movies_df.drop('Genres', axis = 1, inplace = True)

# Create one merged DataFrame containing all the movielens data.
movielens18 = ratings_df.merge(movies_df, on='movie_id')

I'll also take a look at the movies and ratings dataframes.

In [0]:
movies_df.shape

(9742, 3)

In [0]:
movielens18.head()

user_id  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        5  ...  Adventure|Animation|Children|Comedy|Fantasy
2        7  ...  Adventure|Animation|Children|Comedy|Fantasy
3       15  ...  Adventure|Animation|Children|Comedy|Fantasy
4       17  ...  Adventure|Animation|Children|Comedy|Fantasy

[5 rows x 6 columns]

In [0]:
R_df = ratings_df.pivot(index = 'user_id', columns ='movie_id', values = 'rating').fillna(0)
R_df.head()

movie_id  1       2       3       4       ...  193583  193585  193587  193609
user_id                                   ...                                
1            4.0     0.0     4.0     0.0  ...     0.0     0.0     0.0     0.0
2            0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
3            0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
4            0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
5            4.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0

[5 rows x 9724 columns]

de-mean the data (normalize by each users mean) and convert it from a dataframe to a numpy array.

In [0]:
R = R_df.as_matrix()

Z = R>0
m, n = R.shape
Ymean = np.zeros(m)
Ynorm = np.zeros(R.shape)

for i in range(m):
    idx = Z[i, :] == 1
    Ymean[i] = np.mean(R[i, idx])
    Ynorm[i, idx] = R[i, idx] - Ymean[i]    

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [0]:
R_demeaned = R - Ymean[:,np.newaxis]

## Singular Value Decomposition

Scipy and Numpy both have functions to do the singular value decomposition. I'm going to use the Scipy function `svds` because it let's me choose how many latent factors I want to use to approximate the original ratings matrix (instead of having to truncate it after).

In [0]:
from scipy.sparse.linalg import svds
K = 10
U, sigma, Vt = svds(R_demeaned, k = 10)

In [0]:
df.head()

movie_id  ...                                       genres
1         1  ...  Adventure|Animation|Children|Comedy|Fantasy
2         2  ...                   Adventure|Children|Fantasy
3         3  ...                               Comedy|Romance
4         4  ...                         Comedy|Drama|Romance
5         5  ...                                       Comedy

[5 rows x 3 columns]

Done. The function returns exactly what I detailed earlier in this post, except that the $\Sigma$ returned is just the values instead of a diagonal matrix. This is useful, but since I'm going to leverage matrix multiplication to get predictions I'll convert it to the diagonal matrix form.

In [0]:
sigma = np.diag(sigma)

## Making Predictions from the Decomposed Matrices

In [0]:
pd.DataFrame(np.dot(np.dot(U, sigma), Vt)).head() + Ymean[:,np.newaxis]

In [0]:
pd.DataFrame(np.dot(np.dot(U, sigma), Vt)).head()

0         1         2     ...      9721      9722      9723
0  2.731634  0.824226  0.863543  ... -0.123168 -0.123168 -0.133069
1  0.178901 -0.038069 -0.037811  ... -0.002066 -0.002066  0.005347
2  0.023583 -0.001087  0.009023  ... -0.001213 -0.001213 -0.003140
3  1.479939  0.196467  0.192636  ... -0.096388 -0.096388 -0.105531
4  1.256434  0.974787  0.403596  ... -0.020162 -0.020162 -0.020979

[5 rows x 9724 columns]

In [0]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) + preds_df

### Making Movie Recommendations

In [0]:
Ymean[:,np.newaxis].shape

(610, 1)

In [0]:
preds_df = pd.DataFrame(all_user_predicted_ratings, columns = R_df.columns)
print(preds_df.shape)

(610, 9724)


In [0]:
def recommend_movies(predictions_df, userID, movies_df, original_ratings_df, num_recommendations=5):
    
    # Get and sort the user's predictions
    user_row_number = userID -1
    sorted_user_predictions = predictions_df.iloc[user_row_number].sort_values(ascending=False)

    # Get the user's data and merge in the movie information.
    user_data = original_ratings_df[original_ratings_df.user_id == userID]
    user_full = (user_data.merge(movies_df, how = 'left', left_on = 'movie_id', right_on = 'movie_id').
                     sort_values(['rating'], ascending=False))

    print ('User {0} has already rated {1} movies.'.format(userID, user_full.shape[0]))
    print ('Recommending highest {0} predicted ratings movies not already rated.'.format(num_recommendations))
    
    # Recommend the highest predicted rating movies that the user hasn't seen yet.
    recommendations = (movies_df[~movies_df['movie_id'].isin(user_full['movie_id'])]. #all the movies not in the user_full recommendations
         merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
               left_on = 'movie_id',
               right_on = 'movie_id').
         rename(columns = {user_row_number: 'Predictions'}).
         sort_values('Predictions', ascending = False).
                       iloc[:num_recommendations, :-1]
                      )

    return user_full, recommendations

In [0]:
already_rated, predictions = recommend_movies(preds_df, 25, movies_df, ratings_df, 10)

In [0]:
already_rated.head(10)

user_id  ...                                           genres
13       25  ...                                 Action|Drama|War
12       25  ...      Adventure|Animation|Children|Romance|Sci-Fi
23       25  ...                                            Drama
22       25  ...                                      Crime|Drama
19       25  ...                          Action|Adventure|Sci-Fi
18       25  ...                               Drama|Thriller|War
17       25  ...                      Action|Adventure|Crime|IMAX
16       25  ...  Action|Crime|Drama|Mystery|Sci-Fi|Thriller|IMAX
14       25  ...               Adventure|Animation|Children|Drama
1        25  ...                          Action|Adventure|Sci-Fi

[10 rows x 6 columns]

In [0]:
predictions

movie_id  ...                                             genres
275        318  ...                                        Crime|Drama
312        356  ...                           Comedy|Drama|Romance|War
2220      2959  ...                        Action|Crime|Drama|Thriller
895       1196  ...                            Action|Adventure|Sci-Fi
907       1210  ...                            Action|Adventure|Sci-Fi
507        593  ...                              Crime|Horror|Thriller
656        858  ...                                        Crime|Drama
255        296  ...                        Comedy|Crime|Drama|Thriller
3187      4306  ...  Adventure|Animation|Children|Comedy|Fantasy|Ro...
8358    109487  ...                                        Sci-Fi|IMAX

[10 rows x 3 columns]

###Train Test Split

In [0]:
movielens18.shape

(100836, 6)

In [0]:
train_df = movielens18.sample(frac=0.9, random_state=0)
test_df = movielens18.drop(train_df.index.tolist())
train_df.shape, test_df.shape

((90752, 6), (10084, 6))

In [0]:
from sklearn.metrics import mean_absolute_error

# df = fetch_ml20m_ratings()
# movielens18.drop(columns = ['title', 'genres'], inplace=True)
# movielens18.columns = df.columns

train = movielens18.sample(frac=0.8, random_state=7)
val = movielens18.drop(train.index.tolist()).sample(frac=0.5, random_state=8)
test = movielens18.drop(train.index.tolist()).drop(val.index.tolist())

svd = SVD(learning_rate=0.01, regularization=0.1, n_epochs=100,
          n_factors=80, min_rating=1, max_rating=5)

svd.fit(X=train, X_val=val, early_stopping=True, shuffle=False)

pred = svd.predict(test)
mae = mean_absolute_error(test["rating"], pred)


In [0]:
min_user_clicks = 10
filter_users = train_df['user_id'].value_counts() > min_user_clicks
filter_users = filter_users[filter_users].index.tolist()

min_item_clicks = 10
filter_items = train_df['movie_id'].value_counts() > min_item_clicks
filter_items = filter_items[filter_items].index.tolist()

train_filtered = train_df[(train_df['user_id'].isin(filter_users)) & train_df['movie_id'].isin(filter_items)]
print('The original data frame shape:\t{}'.format(train_df.shape))
print('The new data frame shape:\t{}'.format(train_filtered.shape))


In [0]:
#Check that all users appears in both Train & Test

train_users_idx = list(train_filtered.user_id.unique())
test_df_idx = list(test_df.user_id.unique())
iters = list(set(train_users_idx) & set(test_df_idx))


train_filtered = train_filtered[train_filtered.user_id.isin(iters)]
test_df = test_df[test_df.user_id.isin(iters)]

In [0]:
#Check that all movies in Test appears in Train

train_movies_inx = list(train_filtered.movie_id.unique())
test_movies_inx = list(test_df.movie_id.unique())
uniques_movies_test = list(np.setdiff1d(test_movies_inx, train_movies_inx))
flt_lst = list(np.setdiff1d(test_movies_inx, uniques_movies_test))


test_df = test_df[test_df['movie_id'].isin(flt_lst)]

In [0]:
train_users = list(train_filtered.user_id.unique())
test_users = list(test_df.user_id.unique())
counter=0
for user in test_users:
    if user not in train_users:
        counter =+ 1
print("Number of non overlaps in users = {}".format(counter))

train_movies = list(train_filtered.movie_id.unique())
test_movies = list(test_df.movie_id.unique())
counter=0
for movie in test_movies:
    if movie not in train_movies:
        counter =counter + 1

print("Number movies in test that are not in train = {}".format(counter))

In [0]:
pivot_train = train_filtered.pivot_table(values = 'rating', index = 'movie_id', columns = 'user_id')
pivot_train.fillna(0, inplace = True)

pivot_test = test_df.pivot_table(values = 'rating', index = 'movie_id', columns = 'user_id')
pivot_test.fillna(0, inplace = True)

In [0]:
print("Train set contains {} ratings".format(np.sum(np.sum(pivot_train>0))))
print("Test set contains {} ratings".format(np.sum(np.sum(pivot_test>0))))

In [0]:
#Checking if movies in the test set are included in the train set

s_test = set(pivot_test.columns)
s_train = set(pivot_train.columns)

inter = s_test.issubset(s_train)
inter, len(s_train), len(s_test)

### Training the algorithm on the train set

In [0]:
R_df = pivot_train.copy()
R_df.head()

user_id   1    2    3    4    5    6    7    ...  603  604  605  606  607  608  610
movie_id                                     ...                                   
1         4.0  0.0  0.0  0.0  4.0  0.0  4.5  ...  4.0  3.0  4.0  2.5  0.0  2.5  5.0
2         0.0  0.0  0.0  0.0  0.0  4.0  0.0  ...  0.0  0.0  3.5  0.0  0.0  2.0  0.0
3         0.0  0.0  0.0  0.0  0.0  5.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  2.0  0.0
5         0.0  0.0  0.0  0.0  0.0  5.0  0.0  ...  0.0  3.0  0.0  0.0  0.0  0.0  0.0
6         4.0  0.0  0.0  0.0  0.0  4.0  0.0  ...  4.0  3.0  0.0  0.0  0.0  0.0  5.0

[5 rows x 595 columns]

In [0]:
R = R_df.as_matrix()

Z = R>0
m, n = R.shape
Ymean = np.zeros(m)
Ynorm = np.zeros(R.shape)

for i in range(m):
    idx = Z[i, :] == 1
    Ymean[i] = np.mean(R[i, idx])
    Ynorm[i, idx] = R[i, idx] - Ymean[i]    

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [0]:
R_demeaned = R - Ymean[:,np.newaxis]

In [0]:
from scipy.sparse.linalg import svds

In [0]:
'''
The best way I've found to do this..
Looping over the test pivot and locating the corresponding cell in the train set,
by INDEX, where the test-pivot has a rating.
'''
def accuracy(pivot_test, preds_df, k):
    true_val = []
    prediction_val = []

    movie_intersect = list(np.setdiff1d(pivot_test.index, preds_df.index))
    user_intersect = list(np.setdiff1d(pivot_test.columns, preds_df.columns))

    for i in pivot_test.index: #movie_id
        for j in pivot_test.columns: #user_id
            if (pivot_test.loc[i,j]!=0) and (i not in movie_intersect):
                real_value = pivot_test.loc[i,j]
                prediction = preds_df.loc[i,j]
#                 print('movie_id = {},user_id = {}'.format(i,j))
#                 print("real value = {}, prediction = {}".format(real_value, prediction))
                prediction_val.append(preds_df.loc[i,j] + Ymean[:,np.newaxis])
                true_val.append(pivot_test.loc[i,j])

    true_val = np.array(true_val)
    validation_val = np.array(prediction_val)
    RMSE = np.round(np.sqrt(np.square(true_val - prediction_val).mean()),4)
    MAE = np.round(np.abs(true_val - prediction_val).mean(),4)

    print('RMSE: {}, MAE: {}, K: {}'.format(RMSE, MAE, k))
    return (RMSE, MAE)

In [0]:
results = pd.DataFrame(columns=['RMSE', 'MAE','K'])

In [0]:
pivot_train

In [0]:
preds_df.head()

user_id       1         2         3    ...       607       608       610
1        3.451891  0.058611  0.007331  ...  1.408496  1.771635  1.495228
2        0.894914 -0.092731 -0.071086  ...  0.817510  2.109111  0.809867
3        0.835231  0.176799  0.135655  ...  0.291600  2.158124 -1.147367
4       -0.005933  0.141640  0.076914  ... -0.234083  0.370818 -0.210490
5        1.057611 -0.054667  0.036791  ...  0.941970  3.116881  2.827764

[5 rows x 595 columns]

In [0]:
U, sigma, Vt = svds(R_demeaned, k = 50, maxiter= 500)
sigma = np.diag(sigma)
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) + Ymean[:,np.newaxis]
preds_df = pd.DataFrame(all_user_predicted_ratings, columns = R_df.columns)
preds_df.index +=1
# RMSE, MAE = accuracy(pivot_test, preds_df, K)


In [0]:
# flattened_preds = pd.DataFrame(preds_df.to_records())
preds_df['movies'] = preds_df.index
preds_melted = pd.melt(preds_df, id_vars=['movies'], value_vars=preds_df.columns[:595])

In [0]:
preds_melted.head()
preds_melted.columns = ['movie_id', 'user_id', 'rating_predicted']

In [0]:
test_df = pd.merge(test_df, preds_melted,  how='inner', left_on=['user_id','movie_id'], right_on = ['user_id','movie_id'],)


In [0]:
test_df.shape

(3717, 9)

In [0]:
sum(test_df.rating_predicted.isna())

0

In [0]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(test_df.rating, test_df.rating_predicted)

3.1374667166646915

In [0]:
for K in range(1,2):
    U, sigma, Vt = svds(R_demeaned, k = K, maxiter= 500)
    sigma = np.diag(sigma)
    all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) + Ymean[:,np.newaxis]
    preds_df = pd.DataFrame(all_user_predicted_ratings, columns = R_df.columns)
    RMSE, MAE = accuracy(pivot_test, preds_df, K)
    results.append([RMSE, MAE, K])

**The results are pretty bad (too bad, random model givs 1.5 RMSE)..**

We think there's a normalization factor we didn't account for, since the recomendations are okay.